Create Zone -- Portal changes

Sokitha-21 opened the pull request:
Fixes https://github.com/vinyldns/vinyldns/issues/1455.

prerequisites PR - https://github.com/vinyldns/vinyldns/pull/1437

To test the PR - merge the above branch into this and test.

Changes in this pull request:

Create a New Zone from the portal
View, delete, update the generated zone from the portal

nspadaccino committed:
generate zone route

Jay07GIT committed:
created api for zone create with validations

Jay07GIT committed:
Added create zone api in config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Added bind validations

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
json request formatted

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added api key config for zone creation

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
formatted powerdns create zone response

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added create zone api in bind test docker image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added db validations

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
insert values in generate zone db

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added duplicate validation on generate zones

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added pdns config for connect zone

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated generate zones config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
removed unwanted files

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added created timestamp

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added created timestamp test

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add pdns in quickstart

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update pdns config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update pdns config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added list created zones api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added email field in create zone api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added access level field in create zone api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
use json4s to build zone generate requests

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve failed tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve failed tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated generate_zones_bind_api.py

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated manage_vinyldns_zones_bind_api.py with delete zones api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated conf with delete bind uri

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

nspadaccino committed:
zone create provider templates in config

Jay07GIT committed:
updated the get generated zone

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added list generated zone tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

nspadaccino committed:
updated zone generation input and protobuf, added required fields to config

nspadaccino committed:
updates for provider-agnostic zone generation

nspadaccino committed:
update comments

nspadaccino committed:
prune unused optional params from template, enhanced validation and error handling

nspadaccino committed:
Use json schema for validating provider api requests

nspadaccino committed:
convert response message from json string to json

nspadaccino committed:
updated template engine

nspadaccino committed:
add delete and update endpoints, handler updates, added endpoint string substitution

nspadaccino committed:
update zone generation handler, repository methods

nspadaccino committed:
added delete and update generated zone

Jay07GIT committed:
removed the unused commented code

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
upgraded the slf4j-api and logback-classic version to latest

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
updating from logback to log4j2 due to version mismatch in slf4j-api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
undo illegal header config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated create zone params

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve existing tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve existing tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added it tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added it tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added get zones by id

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'master' into create-zone-api

nspadaccino committed:
Merge branch 'master' into logs_dependency_update

nspadaccino committed:
Merge remote-tracking branch 'jay/logs_dependency_update' into create-zone-api

nspadaccino committed:
updated dependencies

Jay07GIT committed:
added functional tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
added functional tests for update,get,delete,list zones

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update func docker network

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'version-0.22.0-beta.1' into create-zone-api

Jay07GIT committed:
trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Merge branch 'version-0.22.0-beta.1' into create-zone-api

Jay07GIT committed:
resolve tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Merge branch 'create-zone-api' of https://github.com/nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

changed the pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

changed the pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

update for pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

update for pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Sokitha-21 committed:
create zone portal changes

Signed-off-by: Sokitha-21 <krishnansokitha@gmail.com>

Sokitha-21 committed:
update

Signed-off-by: Sokitha-21 <krishnansokitha@gmail.com>

Jay07GIT committed:
added deleted zone for bind

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Sokitha-21 committed:
snyk issue fix

Signed-off-by: Sokitha-21 <krishnansokitha@gmail.com>

Sokitha-21 committed:
snyk issue fix

Signed-off-by: Sokitha-21 <krishnansokitha@gmail.com>

nspadaccino committed:
Merge branch 'main' into create-zone-api

nspadaccino committed:
fix typo

nspadaccino committed:
fix typo

nspadaccino committed:
Merge branch 'main' into create-zone-api

JoshSEdwards commented on modules/portal/conf/routes:
```diff
@@ -38,6 +41,11 @@ PUT           /api/zones/:id                     @controllers.VinylDNS.updateZon
 POST          /api/zones/:id/sync                @controllers.VinylDNS.syncZone(id: String)
 
+GET           /api/zones/generate/nameservers    @controllers.VinylDNS.getNameservers
+POST          /api/zones/generate                @controllers.VinylDNS.generateZone
```
The UI submits `PUT /api/zones/generate` in `/modules/portal/public/lib/services/zones/service.zones.js:141` but we're only routing GET, POST, and DELETE for generated zones. I think we need a `PUT /api/zones/generate` route and a proxy method, otherwise the update path will 404.

JoshSEdwards commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -82,6 +165,117 @@ angular.module('controller.zones', [])
+            $scope.updateZoneTemplate = updateZone;
+            $scope.isEditMode = true;
+
+            $scope.zoneFields = Object.entries(createZoneFields).map(([label, config]) => {
```
This is the update zone flow, yet we're displaying `createZoneFields` instead of `updateZoneFields` to the user. Should this be changes to `$scope.zoneFields = Object.entries(updateZone).map(([label, config])`?

JoshSEdwards commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -82,6 +165,117 @@ angular.module('controller.zones', [])
+    };
+
+    $scope.isDynamicZoneFieldRequired = function(fieldName) {
+      return $scope.requiredFields && $scope.requiredFields.indexOf(fieldName) !== -1;
```
The zone creation modal `modules/portal/public/templates/zonecreation-modal.html` uses this method for both the update and create paths, but the required-field lookup is pinned to the create fields `$scope.requiredFields`. This should be made mode-aware and read `$scope.UpdateRequiredFields`. Right now the update required-field list is loaded but never used anywhere.

JoshSEdwards commented on modules/portal/conf/application.conf:
```diff
@@ -86,6 +86,71 @@ api {
     zone-routing-default-max-items = 100
     zone-routing-max-items-limit = 100
   }
+  dns-provider-portal-fields = {
```
Can we avoid re-declaring provider metadata in the portal here? #1437 already establishes this contract on the API side (`modules/api/src/main/resources/application.conf:42`). Is it needed in both application.conf?

Sokitha-21 replied:
Yes, both are needed. The API config holds backend substitution templates and endpoints for calling DNS providers whereas the portal config holds UI field descriptors for rendering form fields. There's no shared config mechanism between the two Play/Akka processes at runtime, so they can't reference each other's config. The provider names overlap, but the structure and purpose are entirely different.

JoshSEdwards commented on modules/portal/public/lib/controllers/controller.zones.js:
This needs to rewire `generateZone-search-text` as well, otherwise you'll check Search by Admin Group on the Created Zones tab, and autocomplete will still show zone names instead of group names.

JoshSEdwards commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -312,6 +578,93 @@ angular.module('controller.zones', [])
             });
     };
 
+    $scope.getUpdatePayload = function (zone) {
```
Similar to my comment on #1437 about separating generated zones with clear boundaries, I think this controller file has gotten way too large for the size of the feature. This file alone almost doubled in length, and this duplicate `getUpdatePayload()` definition (there's another on line 212) is a good signal that this controller is too crowded to keep track of. I think we should probably split generated-zone behavior into its own module/controller.

JoshSEdwards reviewed (changes requested):
Found a few bugs, and I think a refactor to go along with the one in #1437 would be a good idea. Generated zones is a meaty enough feature that it deserves to live in its own files rather than entangled with existing zone code.

Sokitha-21 commented on modules/portal/public/lib/services/zones/service.zones.js:
```diff
@@ -79,17 +79,75 @@ angular.module('service.zones', [])
+        }
+
+        this.getCreateZoneTemplate = function(provider) {
+            var url = "config/createZoneTemplate/" + provider;
```
The URL appears to be missing a leading /.

Sokitha-21 commented on modules/portal/conf/routes:
```diff
@@ -17,6 +17,9 @@ GET           /dnschanges                        @controllers.FrontendController
 GET           /dnschanges/:id                    @controllers.FrontendController.viewBatchChange(id: String)
 GET           /recordsets                        @controllers.FrontendController.viewRecordSets
+GET           /config/allowedDNSProviders        @controllers.FrontendController.getAllowedDNSProviders
+GET           /config/createZoneTemplate/:provider   controllers.FrontendController.getCreateZoneTemplate(provider: String)
```
This route is missing the @ prefix before controllers.FrontendController

Sokitha-21 commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -312,6 +578,93 @@ angular.module('controller.zones', [])
+    };
+
+    $scope.deleteCreatedZone = function () {
+        if (!confirm("Are you sure you want to delete this zone?")) return;
```
Browser `confirm()` is not recommended as it blocks the event loop. Suggest to use an angular modal pattern instead.

Sokitha-21 commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -59,13 +72,83 @@ angular.module('controller.zones', [])
+    $scope.viewCreatedZone = function (createdZone) {
+        $log.debug(createdZone)
+        $scope.createZone = angular.copy(createdZone);
+        $scope.isEditMode = false;
```
`viewCreatedZone` resets `isEditMode` but never clears `$scope.isDelete ` unlike `updateCreatedInfo`. As a result, `isDelete` may retain its previous value when opening a zone via `viewCreatedZone` causing the delete button to be shown or hidden incorrectly.

Sokitha-21 commented on modules/portal/public/lib/controllers/controller.zones.js:
```diff
@@ -312,6 +578,93 @@ angular.module('controller.zones', [])
+            const payload = $scope.getUpdatePayload($scope.createZone);
+            zonesService.updateGeneratedZone(payload)
+              .then(function (response) {
+                $("#zone_creation_modal").modal("hide");
```
The create path inside `addZoneCreation` calls `$timeout($scope.refreshGeneratedZones(), 1000)` after success.The update (else) path is missing that same refreshGeneratedZones() call, so the edited zone's data stays stale in the table after saving.

Jay07GIT committed:
resolved the review comments

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated the sql syntax

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolved func tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added auth tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added auth tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
security: enforce access control on generated-zone list endpoints

  listGenerateZones and listGeneratedZonesByAdminGroupIds returned every
  generated zone in the system regardless of the caller's group membership:
  the repository methods took no AuthPrincipal and never applied an access
  filter, so the `ignoreAccess` flag was inert and any authenticated user
  could enumerate all generated zones.

  Generated zones have no per-zone ACL table (unlike regular zones), so
  access is purely admin-group membership. Add an accessFilter helper that
  mirrors MySqlZoneRepository.withAccessors semantics: super/support users,
  or callers explicitly passing ignoreAccess, see all zones; everyone else
  is restricted to zones owned by a group they belong to; a user in no
  groups sees nothing. Thread AuthPrincipal through the repository algebra
  and ZoneService into both list queries.

  Add integration regression tests covering: non-members cannot see another
  group's zones (direct and admin-group-search paths), members and super
  users can, and ignoreAccess restores the unfiltered view.

nspadaccino committed:
security: encrypt provider API key instead of holding it in plaintext

  DnsProviderConfig.apiKey was a plain String loaded from config and passed
  around in memory, diverging from how every other connection secret in
  VinylDNS is handled (Encrypted via CryptoAlgebra).

  Change apiKey to Encrypted, encrypt it at config-load time, and decrypt
  it just-in-time at the three ZoneService call sites only when setting the
  X-API-Key header. The in-memory representation is now ciphertext, so the
  key cannot leak through accidental logging or toString.

nspadaccino committed:
security: drop server-owned fields from generate-zone request model

  ZoneGenerationInput (the generate-zone request body) carried id, status,
  and response with defaults, and the JSON deserializer read them straight
  from the request. A client could therefore set a zone's id/status or
  inject a fabricated provider response, which then flowed into the
  persisted GenerateZone.

  Remove id, status, and response from ZoneGenerationInput and stop reading
  them in ZoneGenerationInputSerializer. The server now owns these fields:
  GenerateZone assigns a fresh id and Active status, and ZoneService sets
  the response from the actual provider call. Add a routing regression test
  that posts id/status in the body and asserts they are ignored.

nspadaccino committed:
security: fail closed when no provider schema is configured

  schemaValidationResult returned success when a provider had no JSON schema
  for an operation, so a provider misconfigured without a create-zone or
  update-zone schema would forward unvalidated providerParams straight to
  the external provider.

  Reject in that case instead. A missing schema is an operator
  misconfiguration rather than bad client input, so surface it as a
  server-side failure (500). Add a service test asserting a provider with
  no schema is rejected before any provider call.

nspadaccino committed:
security: bound external DNS-provider calls with timeouts

  createConnection opened an HttpURLConnection with no connect or read
  timeout, so a slow or hung provider would block the request thread
  indefinitely. Set a 10s connect and 30s read timeout. A configurable
  client abstraction is left for the follow-up refactor.

nspadaccino committed:
test: update generated-zone list functional tests for access scoping

  Phase 0 added access control to the generated-zone list endpoints, so a
  caller now sees only zones owned by groups they belong to. Four functional
  tests in list_generated_zones_test.py asserted counts/pagination from the
  old global-visibility behavior (e.g. has_length(7)/(8)) and failed in CI.

  Update them to the access-scoped 5-zone set owned by list_generate_zones_group,
  matching the already-passing second_page test.

nspadaccino committed:
review: address PR feedback on generated-zone hardening

  Three changes from Josh's review:

  - ignoreAccess: remove the parameter entirely from the generated-zone list
    endpoints instead of leaving it as a bypass. Access is now role-based —
    super/support admins see all generated zones, every other caller is always
    scoped to the groups they belong to. Drops the query param, the repository/
    service params, and the ignoreAccess field from ListGeneratedZonesResults
    and ListGeneratedZonesResponse.

  - providerParams: reject a request whose providerParams is present but not a
    JSON object (400 "providerParams must be a JSON object") instead of quietly
    collapsing the invalid payload into an empty map.

  - access filter: cap the member-group accessor list at MAX_ACCESSORS (30) with
    a warning, mirroring MySqlZoneRepository.buildZoneSearchAccessorList, to
    avoid building an unbounded parameterized query.

  Tests updated accordingly: ZoneServiceSpec/ZoneRoutingSpec/integration specs no
  longer thread ignoreAccess; added a routing test for malformed providerParams;
  functional list tests now assert access-scoped results with admins seeing all.

nspadaccino committed:
Merge pull request #1 from nspadaccino/czapi-phase0-security

Phase 0: security hardening for the generated-zone API

nspadaccino committed:
fix(C2): reject provider switch on generated-zone update

  The update path selected the provider endpoint from request.provider but never
  updated the saved record's provider, so an update could silently send traffic to
  a different provider while leaving the stored record inconsistent. Validate that
  request.provider matches the existing zone's provider and reject the change.

nspadaccino committed:
fix(C3): replace providerParams on generated-zone update

The update path merged existing.providerParams ++ request.providerParams, so a param could never be removed once set. Use standard PUT semantics: the request's providerParams fully replace the stored set.

nspadaccino committed:
fix(C6): checked endpoint lookup and reject unresolved URL placeholders

Bare providerConfig.endpoints(op) Map.apply threw a raw NoSuchElementException on missing operator config, undermining the fail-closed contract that a missing schema already honors. Look endpoints up with a checked helper that fails closed. Also reject endpoint URLs that still contain {{key}} placeholders after substitution rather than sending a malformed URL to the provider (JSON-body placeholder pruning is unchanged).

nspadaccino committed:
fix(C7): cross-table existence check on generate

handleGenerateZoneRequest now also calls zoneDoesNotExist so a generated zone cannot collide with an existing VinylDNS-managed zone. The reverse check in connectToZone is intentionally omitted: the supported two-step flow generates a zone then connects to it by the same name, so an existing generate_zone record must not block connecting to that zone.

nspadaccino committed:
chore: hygiene (H1, H3, H5) in generated-zone paths

  H1: rename shadowing param ConnectZoneInput -> connectZoneInput in
      connectToZone / encrypt (+ scaladoc).
  H3: delete the commented-out generateZoneExists block.
  H5: de-dupe the double canChangeZone in the zone and generated-zone update
      paths - canChangeZone authorizes purely on admin-group membership, so the
      second check now runs only when the admin group actually changes.

nspadaccino committed:
refactor(B3): delete dead generate-zone types

  Remove UpdateGenerateZoneInput and the unused GenerateZone.apply overload that
  consumed it, plus the unused RRSet/Record case classes (only self-referencing).
  Drop the test and fixture that exercised the dead apply overload; the update-zone
  routing mock now returns the GenerateZone fixture directly (previously an
  UpdateGenerateZoneInput cast to GenerateZone).

nspadaccino committed:
refactor(B1): move generate-zone domain types into core.domain.zone.generate

  Pure move, no behavior change. Relocate GenerateZone (+companion), the
  GenerateZoneStatus/GenerateZoneChangeType enums, ZoneGenerationInput,
  ZoneGenerationResponse, ListGeneratedZonesResults, and GenerateZoneRepository
  out of the shared zone package into a dedicated vinyldns.core.domain.zone.generate
  package, and fix imports across core/mysql/api. DnsProviderConfig/
  DnsProviderApiConnection and the provider config parsing stay in Zone.scala for B2.

nspadaccino committed:
refactor(B2): move DNS-provider config + parsing into generate package

  Move DnsProviderConfig/DnsProviderApiConnection and the provider config-parsing
  block out of Zone.scala into generate/DnsProviderApiConnection.scala as
  DnsProviderApiConnection.load(config, crypto). Parsing logic is copied verbatim
  (no pureconfig conversion, so error/missing-key semantics are unchanged);
  ConfiguredDnsConnections.load now just calls the loader. Drop the now-unused
  ConfigFactory import.

nspadaccino committed:
refactor(B4): collapse GenerateZoneStatus into ZoneStatus

  GenerateZoneStatus had byte-identical values (Active/Deleted/Syncing) to
  ZoneStatus, so JSON and protobuf wire formats are unchanged. Replace all uses of
  GenerateZoneStatus with ZoneStatus and delete the enum; drop the now-redundant
  JsonEnumV(GenerateZoneStatus) registration (ZoneStatus is already registered).
  The 'Pending*' -> Active mapping in ProtobufConversions.fromPB is preserved for
  old stored data with statuses outside the current enum.

nspadaccino committed:
refactor(B5): extract GenerateZoneService from ZoneService

  Split the generate-zone concern out of ZoneService into a dedicated
  GenerateZoneService + GenerateZoneServiceAlgebra: the generate handlers, the
  DNS-provider HTTP client (createConnection/createDnsZoneService), TemplateEngine,
  JsonSchemaValidator, listGeneratedZones, and the provider lookup helpers. Wire the
  new service through ZoneRoute, VinylDNSService, and Boot; ZoneService no longer
  depends on dnsProviderApiConnection or the generate repository. Move the generate
  service tests into GenerateZoneServiceSpec; ZoneRoutingSpec's mock now implements
  both algebras. Behavior unchanged.

nspadaccino committed:
test: give generated list-zones fixtures distinct names (C7 fix)

  C7 added a cross-table check so a generated zone can't collide with a managed
  zone. The shared functional fixture created managed zones list-zones-test-searched/
  unfiltered-* and then generated zones with the same names, which C7 now rejects
  (409), breaking shared_zone_test_context setup and cascading to ~620 test errors.
  Rename the generated fixtures to list-generated-zones-test-* so they no longer
  collide; the searched/unfiltered substrings and ordering are preserved, so the
  name-filter and pagination assertions in list_generated_zones_test.py still hold.

nspadaccino committed:
test: fix duplicate/literal group name in create_group tests

  test_create_group_success_wildcard and test_create_group_success_number_of_dots
  both used the identical literal name "...success_wildcard{shared_zone_test_context.partition_id}"
  with no f-string prefix, so partition_id never interpolated and the two tests
  collided across xdist workers (409 'Group ... already exists'). Add the f prefix
  and give each test a distinct, partition-scoped name. Pre-existing bug unrelated
  to the generated-zone changes; surfaced once the C7 fixture fix let the suite run.

Sokitha-21 committed:
Merge remote-tracking branch 'upstream/create-zone-api' into generate-zone-portal-changes

Sokitha-21 committed:
Merge remote-tracking branch 'upstream/create-zone-api-refactor' into generate-zone-portal-changes

Sokitha-21 committed:
resolved comments
